swarm_dial.go

```diff
@@ -281,39 +286,311 @@ func (s *Swarm) dialPeer(ctx context.Context, p peer.ID) (*Conn, error) {
+	defer s.limiter.clearAllPeerDials(p)
 
-	conn, err := s.dial(ctx, p)
-	if err != nil {
-		conn := s.bestAcceptableConnToPeer(ctx, p)
-		if conn != nil {
-			// Hm? What error?
-			// Could have canceled the dial because we received a
-			// connection or some other random reason.
-			// Just ignore the error and return the connection.
-			log.Debugf("ignoring dial error because we already have a connection: %s", err)
-			return conn, nil
+	type pendRequest struct {
+		req   DialRequest               // the original request
+		err   *DialError                // dial error accumulator
+		addrs map[ma.Multiaddr]struct{} // pending addr dials
+	}
+
+	type addrDial struct {
+		ctx      context.Context
+		conn     *Conn
+		err      error
+		requests []int
+	}
+
+	reqno := 0
+	requests := make(map[int]*pendRequest)
+	pending := make(map[ma.Multiaddr]*addrDial)
+
+	var triggerDial <-chan time.Time
+	var nextDial []ma.Multiaddr
+	active := 0
+	done := false
+
+	resch := make(chan dialComplete)
+
+loop:
+	for {
+		select {
+		case req, ok := <-reqch:
+			if !ok {
+				// request channel has been closed, wait for pending dials to complete
+				if active > 0 {
+					done = true
+					reqch = nil
+					triggerDial = nil
+					continue loop
+				}
+
+				// no active dials, we are done
+				return
+			}
+
+			c := s.bestAcceptableConnToPeer(req.Ctx, p)
+			if c != nil {
+				req.Resch <- DialResponse{Conn: c}
+				continue loop
+			}
+
+			addrs, err := s.addrsForDial(req.Ctx, p)
+			if err != nil {
+				req.Resch <- DialResponse{Err: err}
+				continue loop
+			}
+
+			// at this point, len(addrs) > 0 or else it would be error from addrsForDial
+			// ranke them to process in order
+			addrs = s.rankAddrs(addrs)
+
+			// create the pending request object
+			pr := &pendRequest{
+				req:   req,
+				err:   &DialError{Peer: p},
+				addrs: make(map[ma.Multiaddr]struct{}),
+			}
+			for _, a := range addrs {
+				pr.addrs[a] = struct{}{}
+			}
+
+			// check if any of the addrs has been successfully dialed and accumulate
+			// errors from complete dials while collecting new addrs to dial/join
+			var todial []ma.Multiaddr
+			var tojoin []*addrDial
+
+			for _, a := range addrs {
+				ad, ok := pending[a]
+				if !ok {
+					todial = append(todial, a)
+					continue
+				}
+
+				if ad.conn != nil {
+					// dial to this addr was successful, complete the request
+					req.Resch <- DialResponse{Conn: ad.conn}
+					continue loop
+				}
+
+				if ad.err != nil {
+					// dial to this addr errored, accumulate the error
+					pr.err.recordErr(a, ad.err)
+					delete(pr.addrs, a)
+				}
+
+				// dial is still pending, add to the join list
+				tojoin = append(tojoin, ad)
+			}
+
+			if len(todial) == 0 && len(tojoin) == 0 {
+				// all request applicable addrs have been dialed, we must have errored
+				req.Resch <- DialResponse{Err: pr.err}
+				continue loop
+			}
+
+			// the request has some pending or new dials, track it and schedule new dials
+			reqno++
```

Comment:
Not really. We need to pack request ids in the `addrDial` struct and then check to see if they have been complete (removed from the tracking).
If we used the request object as key we'd have to copy the requests all over the place and then do extra management stuff to keep them up-to-date.

What's wrong with using an integer? It doesn't cost anything.

Reply:
It works, it's just something else to track.

>  If we used the request object as key we'd have to copy the requests all over the place and then do extra management stuff to keep them up-to-date.

Can't we use the pointers? No need to copy anything and should be the same size as an int.

Reply:
(obviously not a huge deal)

Reply:
Well, we could use a req pointer but then we'd have to reify the struct and move it to the heap -- which is strictly worse than using integers, because there is one more thing for the gc to track. Not a huge impact though.

I would agree with tracking overhead, but the way we use the integers is really simple,
The reqno is implemented at exactly one place, right before we add to the tracking map, so there is very little mental overhead involved.

I am really inclined to leave it as is, unless you feel very strongly about it.

Reply:
I'm happy to leave it as it is, but I think I'm confused about something. The request is allocated here: 

https://github.com/libp2p/go-libp2p-swarm/pull/250/files#diff-3025a253202bbb3d70750d06692c51f9361993ecafcce2bddbebd5508337faaaR415-R419

and we're storing it by reference in the `requests` map. It should be possible to just change the requests map to `map[*pendRequest]struct{}`. and `ad.requests` to `[]*pendRequest`.

Actually... I _think_ we can even get rid of the `requests` map.

Reply:
We can't that easily, we will have to track what request is still active.
That's the main reason for the indirection, and that's why I like the integers as pointers approach -- it forces you to check for validity when you want to use it.